### PR TITLE
LPAL-724 Use CMK for LB access logs bucket

### PR DESCRIPTION
**Pull request by ntse**
## Purpose

Switch from AWS-managed keys to CMK keys for access log
Fixes LPAL-724

## Approach

Use the recently created CMK KMS keys for the loadbalancer access logs S3 bucket

## Learning

n/a

## Checklist

* [X] I have performed a self-review of my own code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [ ] I have added tests to prove my work
* [ ] I have added mandatory tags to terraformed resources, where possible
* [ ] If I have a new OPG component dependency, I have updated the `metadata.json` with the repo location.
* [ ] If I added a package.json or composer.json, I also made sure this is included in the script in `.github/workflows/dependabot-update.yml`
* [ ] The product team have tested these changes

### CODE REVIEW

File: terraform/region/kms.tf [hunk 1]
```diff
@@ -1,7 +1,79 @@
+data "aws_kms_key" "access_log_key" {
+  key_id = "alias/mrk_access_logs_lb_encryption_key-${terraform.workspace}"
+}
+
+data "aws_iam_policy_document" "s3_loadbalancer_kms" {
```
Comment (townxelliot) [inline terraform/region/kms.tf:5]: Could you please explain why you need to enable root and key administrator permissions? Is it not possible to make a policy which only allows access from the LB (i.e. the third statement)? Or does the policy have wider application beyond the LB?

File: terraform/region/data_sources.tf [hunk 2]
```diff
@@ -0,0 +1 @@
+data "aws_caller_identity" "current" {}
```
Comment (townxelliot) [inline terraform/region/data_sources.tf:1]: What's the purpose of this please?

File: terraform/region/kms.tf [hunk 3]
```diff
@@ -1,7 +1,79 @@
...
+
+  statement {
+    sid    = "Allow ELB to use Key for Encryption"
+    effect = "Allow"
+
+    actions = [
+      "kms:Encrypt",
+      "kms:Decrypt",
+      "kms:ReEncrypt*",
+      "kms:GenerateDataKey*",
```
Comment (townxelliot) [inline terraform/region/kms.tf:56]: Does the ELB need to be able to describe keys and generate data keys? In what context would this happen?

**Review by townxelliot — CHANGES_REQUESTED** [review]
I've added a few requests for clarification. It would probably help the team if you organised a PR party to explain CMKs, why we're using them, how they're scoped etc.

File: terraform/region/data_sources.tf (hunk 2 above)
Comment (williamfalconeruk) [reply terraform/region/data_sources.tf:1]: this provides information to terraform for it's current aws context, which can be referenced by the configuration. See: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity

File: terraform/region/kms.tf (hunk 3 above)
Comment (williamfalconeruk) [reply terraform/region/kms.tf:56]: GenerateDataKey : https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateDataKey.html - is required for the way that keys are generated, as there is a master key (CMK) and a data key which encrypts over the top, outside of KMS. this is usually stored with the data.
DescribeKey - https://docs.aws.amazon.com/kms/latest/APIReference/API_DescribeKey.html - allows the Principal to get information about a key.

**Comment by townxelliot** [comment]
Thanks for clarifications @williamfalconeruk.

**Review by townxelliot — APPROVED** [review]

**Review by williamfalconeruk — APPROVED** [review]

File: terraform/region/kms.tf [hunk 4]
```diff
@@ -15,6 +87,7 @@ resource "aws_kms_replica_key" "multi_region_access_logs_lb_encryption_key_repli
   deletion_window_in_days = 7
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/region/kms.tf:87]: This is to do with ensuring there is a minimum time left prior to actual deletion, as when you ask aws to do so it "marks" as deleted but retains the key for this period of time i.e. 7 days as specified here. Allows a recovery to happen if this was done by mistake.

**Pull request merged**